src/iceberg/util/snapshot_util.cc

```diff
@@ -0,0 +1,373 @@
+}
+
+Result<bool> SnapshotUtil::IsParentAncestorOf(const Table& table, int64_t snapshot_id,
+                                              int64_t ancestor_parent_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, AncestorsOf(table, snapshot_id));
+  for (const auto& snapshot : ancestors) {
+    if (snapshot->parent_snapshot_id.has_value() &&
+        snapshot->parent_snapshot_id.value() == ancestor_parent_snapshot_id) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Result<std::vector<std::shared_ptr<Snapshot>>> SnapshotUtil::CurrentAncestors(
+    const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto current, table.current_snapshot());
+  return AncestorsOf(table, current);
+}
+
+Result<std::vector<int64_t>> SnapshotUtil::CurrentAncestorIds(const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, CurrentAncestors(table));
+  return ToIds(ancestors);
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestor(
+    const Table& table) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, CurrentAncestors(table));
+  if (ancestors.empty()) {
+    return std::nullopt;
+  }
+  return ancestors.back();
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestorOf(
+    const Table& table, int64_t snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto ancestors, AncestorsOf(table, snapshot_id));
+  if (ancestors.empty()) {
+    return std::nullopt;
+  }
+  return ancestors.back();
+}
+
+Result<std::optional<std::shared_ptr<Snapshot>>> SnapshotUtil::OldestAncestorAfter(
+    const Table& table, TimePointMs timestamp_ms) {
+  ICEBERG_ASSIGN_OR_RAISE(auto current, table.current_snapshot());
+  if (!current) {
+    // there are no snapshots or ancestors
+    return std::nullopt;
+  }
+
+  std::shared_ptr<Snapshot> last_snapshot = nullptr;
+  auto ancestors = AncestorsOf(table, current);
+  for (const auto& snapshot : ancestors) {
+    auto snapshot_timestamp_ms = snapshot->timestamp_ms;
+    if (snapshot_timestamp_ms < timestamp_ms) {
+      return last_snapshot ? std::make_optional(last_snapshot) : std::nullopt;
+    } else if (snapshot_timestamp_ms == timestamp_ms) {
+      return snapshot;
+    }
+    last_snapshot = snapshot;
+  }
+
+  if (last_snapshot && !last_snapshot->parent_snapshot_id.has_value()) {
+    // this is the first snapshot in the table, return it
+    return last_snapshot;
+  }
+
+  return ValidationFailed("Cannot find snapshot older than {}",
+                          FormatTimestamp(timestamp_ms));
+}
+
+Result<std::vector<int64_t>> SnapshotUtil::SnapshotIdsBetween(const Table& table,
+                                                              int64_t from_snapshot_id,
+                                                              int64_t to_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto to_snapshot, table.SnapshotById(to_snapshot_id));
+  if (!to_snapshot) {
+    return InvalidArgument("Cannot find snapshot: {}", to_snapshot_id);
```

Comment:
`InvalidArgument` should be used for input argument. This is an invalid state so perhaps return `Invalid`?